packages/model-viewer/src/features/scene-graph/model.ts

```diff
@@ -13,42 +13,275 @@
+  }
+}
+
+// Represents a primitive in a glTF mesh.
+class PrimitiveNode extends Node {
+  private[$mesh]: Mesh;
+  // Maps glTF material index number to a material that this primitive supports.
+  private[$materials] = new Map<number, Material>();
+  // Maps variant name to material index.
+  private[$variantInfo]: Map<string, Material>;
+  // List of child nodes.
+  constructor(
+      mesh: Mesh, mvMaterials: Material[],
+      correlatedSceneGraph: CorrelatedSceneGraph) {
+    super(mesh.name);
+    this[$mesh] = mesh;
+    const {gltf, threeGLTF} = correlatedSceneGraph;
+
+    // TODO: Remove the associationKey 'work arounds' after fixing Three.js
+    // associations. This is needed for now because Three.js does not create
+    // associations with SkinnedMeshes (glTF primitives of an animated object)
+    // and incorrect associations are formed when a mesh has multiple
+    // primitives.
+    let associationKey: Object3D = mesh;
+    // Work around 1, skinned meshes (glTF primitives) have no association but
+    // the parent (mesh) does, which maps to a 'node'.
+    if (mesh instanceof SkinnedMesh && mesh.parent != null) {
+      associationKey = mesh.parent;
+    }
+    let gltfMeshReference = correlatedSceneGraph.threeObjectMap.get(
+                                associationKey) as GLTFReference;
+    // Work around 2, the association value is a material not a node.
+    if (gltfMeshReference.type === 'materials') {
+      associationKey = mesh.parent!;
+      gltfMeshReference = correlatedSceneGraph.threeObjectMap.get(
+                              associationKey) as GLTFReference;
+    }
+
+    const {type: nodes, index: nodeIndex} = gltfMeshReference;
+
+    // Should have the correct reference type now.
+    if (nodes !== 'nodes') {
+      console.error('Expected type \'nodes\' but got ' + nodes);
+    }
+    // Gets the mesh index from the node.
+    const meshIndex = ((gltf[nodes] || []) as GLTFNode[])[nodeIndex].mesh!;
+    // The gltf mesh array to sample from.
+    const meshElementArray = gltf['meshes'] || [];
+    // List of primitives under the mesh.
+    const gltfPrimitives =
+        (meshElementArray[meshIndex].primitives || []) as Primitive[];
+
+    for (const primitive of gltfPrimitives) {
+      // Maps the primitive default to a material.
+      this[$materials].set(
+          primitive.material!, mvMaterials[primitive.material!]);
+
+      if (primitive.extensions &&
+          primitive.extensions['KHR_materials_variants']) {
+        const variantsExtension =
+            primitive.extensions['KHR_materials_variants'] as
+            KHRMaterialsVariants;
+        const extensions = threeGLTF.parser.json.extensions;
+        const variantNames = extensions['KHR_materials_variants'].variants;
+        // Provides definition now that we know there are variants to support.
+        this[$variantInfo] = new Map<string, Material>();
+        for (const mapping of variantsExtension.mappings!) {
```

Comment:
Well, the fact that we have a variantsExtension to start with is a pretty good indication that mappings exist unless the file is malformed.

Reply:
If that's true I  would expect variantsExtension.mappings to be a type  that doesn't include `null`/`undefined`. Maybe  just needs an update to the type? However, I think mappings can be absent for a mesh that has the same material in all variants?

Reply:
The spec does not definitively say one way or the other but looking at the implementation it seems clear:
If mappings is not defined the variant is malformed, I'm going to remove the undefined option from our interface.
```
"primitives": [
      {
        ...,
        "extensions": {
          "KHR_materials_variants" : {
            "mappings": [
              {
                "material": 2,
                "variants": [0, 3],
              },
              {
                "material": 4,
                "variants": [1],
              },
              {
                "material": 5,
                "variants": [2],
              },
            ],
          }
        }
      },
      // ... more primitives ...
    ]
```

Reply:
Hmm, I don't think that's true. From the spec:

```
Across the entire mappings array, each variant index must be used no more than one time.

When the active variant is referenced in a mapping, a compliant viewer will apply its material to the primitive.

When no mapping contains the active variant, or there is no active variant, a compliant viewer will fall back on vanilla glTF behaviour for the primitive.
```

I think when there's no mapping, you use the initially-loaded (non-extension) material. I think handling this fits into handing the return to initial case.

Reply:
I think we are saying the same thing overall, I'm just trying to make the point that the spec does not make room for having a variant definition in a primitive that has an undefined `mappings` field.
like:
```
"extensions": {
          "KHR_materials_variants" : {
             // Can't have just emptiness here
            ],
          }
        }
```
I think there is a problem with the `return to initial case`
Say you have model that support variants `red` and `blue` for instance, where `red` is the default/initial material.
Then you call `model.switchVariant(blue)` it switches to blue, then you call `model.switchVariant(green)` then it switches to... `red`? This would be because technically there is no mapping for `green` so it defaults back to the initial material. But the spec says to fall back on the vanilla glTF _behavior_ not fall back to the initial material, which they could have easily written, the default _behavior_ does not allow switching materials at all so in other words: just ignore it.